Power supply package

Sharpe49 opened the pull request:
Including :
- Master key
- Battery switch
- Service retention
- Cabview controls shown only when battery power supply or cab power supply is switched on
- Electric train supply
- Passenger car power supply and HVAC
- Diesel power supply (with power consumption from electric train supply)
- Dual mode power supply (in order to implement dual mode locomotives later)
- Traction cut-off relay
- TCS additions linked to power supplies

Feature request: Missing (please add Elvas Tower, Trello, Launchpad link if one exists)

Sharpe49 committed:
Power supply package
Including :
- Master key
- Battery switch
- Service retention
- Cabview controls shown only when battery power supply or cab power supply is switched on
- Electric train supply
- Passenger car power supply and HVAC
- Diesel power supply (with power consumption from electric train supply)
- Dual mode power supply (in order to implement dual mode locomotives later)
- Traction cut-off relay
- TCS additions linked to power supplies

YoRyan commented:
Woah, this is a massive PR. Does this represent all of the power supply additions in the NewYear fork? I think it's too large for a single person to review.

By comparison, the Monogame upgrade, #40, was abut one-fifth the size, and it was rejected by @twpol for being overly complex and containing too many unexplained changes.

Csantucci commented:
To have some test I locally tried to merge the PR into a branch derived from the actualized master branch, but I got many rejected hunks, all related to .pot files. Not important to have tests, but I wanted to share this to you.

Sharpe49 commented:
> To have some test I locally tried to merge the PR into a branch derived from the actualized master branch, but I got many rejected hunks, all related to .po or .pot files. Not important to have tests, but I wanted to share this to you. Probably you forked from the master branch before the latest changes to the locals files by Weter.

The parent commit is https://github.com/openrails/openrails/commit/f497b6b6da69357020c92a9e214b379c5d98f792
which is the latest commit on the master branch (I always rebase my branch on the master branch before pushing).
I regenerated the POT files in my working directory, no modifications were created. So your branch is probably the one having some problems with the translation files (probably not up to date).

Sharpe49 commented:
> Woah, this is a massive PR. Does this represent all of the power supply additions in the NewYear fork? I think it's too large for a single person to review.
> 
> By comparison, the Monogame upgrade, #40, was abut one-fifth the size, and it was rejected by @twpol for being overly complex and containing too many unexplained changes.

I know, but many of the added functionalities are interlinked. I makes no sense separating them. I could probably create 2, 3 or 4 separate commits, but it would take a long time to separate them logically and the commits would still be quite huge.

cesarBLG commented:
I've had a quick look at the code and, although introducing a considerable amount of new code, it's not complex. Also, not taking into account locales, documentation or moved code, the size of the commit is reduced more than half. There's a lot of initialization and save/restore code, which doesn't need a deep review.

Separating the different traction subsystems from the Electric/Diesel locomotive classes is necessary to correctly reproduce some locomotives (for example, dual locomotives, electric trains without pantographs or with gearboxes), and I believe is in line with [James' architecture draft](https://github.com/twpol/openrails/blob/feature/organise-github-files/Docs/Architecture.md).

YoRyan commented:
This PR *does* merge cleanly, but the code bot is excluding it from the unstable build, possibly because of a conflict with #389 (which itself is currently failing CI).

cesarBLG commented:
> This PR _does_ merge cleanly, but the code bot is excluding it from the unstable build, possibly because of a conflict with #389 (which itself is currently failing CI).

It might be something else, because the Code Bot doesn't have Weter in the list of members, so that PR has to be explicitly included in the unstable branch.

Csantucci commented:
1) In Train.cs, starting from line 167, there are some lines defining some variables cancelled, like e.g. 
![immagine](https://user-images.githubusercontent.com/46676244/115903214-9c8aed00-a463-11eb-9ade-ab5b4113a38e.png)
Is this wanted, and where have been moved such variables?
2) Some user commands added use special keys, like Insert and Enter: Can't this have drawbacks?
3) What must added/modified to implement dual mode locomotives?
4) Is dual mode general enough to allow also for other dual modes (e.g. pantograph/battery powered)?

Sharpe49 committed:
Added HandleEventFromLeadLocomotive functions

Sharpe49 commented:
> > This PR _does_ merge cleanly, but the code bot is excluding it from the unstable build, possibly because of a conflict with #389 (which itself is currently failing CI).
> 
> It might be something else, because the Code Bot doesn't have Weter in the list of members, so that PR has to be explicitly included in the unstable branch.
> 
> The console output from James website shows that there are conflicts with CabViewFile.cs and TrainControlSystem.cs, so probably it's conflicting with #384 (and maybe #391).

I just sent a rebase from master. This should be fixed.
I also added a commit including some small modifications I made this week (two new HandleEvent functions used when other cars are receiving orders from the lead locomotive).
I will also push to my script repository some classes that shows the changes of behaviour of the power supply according to the French rules.

> 1. In Train.cs, starting from line 167, some lines defining some variables have been deleted, like e.g.
>    public float TrainSteamHeatLossWpT;             // Total Steam Heat loss of train
>    Is this wanted, and where have been such variables moved to?
> 2. Some user commands added use special keys, like Insert and Enter: Can't this have drawbacks?
> 3. What must still be added/modified to implement dual mode locomotives?
> 4. Is dual mode general enough to allow also for other dual modes (e.g. pantograph/battery powered)?

1. They are removed because they are unused. Most variables used in the heating calculations are located in TrainCar.
2. I don't think so.
3. Basically, creating a new locomotive type able to store parameters for both electric and diesel traction. I think, we should wait for the JSON format for cars before implementing this.
4. Currently, no. But this is indeed something to think about when we will create dual mode locomotives.

Csantucci commented:
Thanks for replies. This crash http://www.elvastower.com/forums/index.php?/topic/35128-error-with-or-u20210425-0921/page__pid__271465#entry271465 is caused by following code line modified by this PR
`                        if (!LocomotivePowerSupply.MainPowerSupplyOn)`
and related to an AI train and to a steam locomotive.

Sharpe49 commented:
> Thanks for replies. This crash http://www.elvastower.com/forums/index.php?/topic/35128-error-with-or-u20210425-0921/page__pid__271465#entry271465 is caused by following code line modified by this PR
> ` if (!LocomotivePowerSupply.MainPowerSupplyOn)`
> and related to a steam locomotive of an AI train. LocomotivePowerSupply is null.

It should be fixed with latest commit.

Sharpe49 committed:
Fixed Copy, Save and Restore functions in MasterKey and BatterySwitch

Sharpe49 commented:
Ok, since there are a lot of variables coming from the power supply in MSTSLocomotive, instead of null checking everything in the case of the steam locomotive, I will create a stub (a SteamPowerSupply class that returns always the same values => PowerOn).

Sharpe49 committed:
Fixes in order to make steam locomotives work with the power supply package
(and add a basic electric power supply for electric systems in steam locomotives)

Sharpe49 committed:
Update locales

Sharpe49 committed:
Added DisabledIfCabPowerSupplyOff and DisabledIfLowVoltagePowerSupplyOff parameters to ETCS DMI

Sharpe49 committed:
Dynamic brake availability is now managed by power supply scripts
(related to https://bugs.launchpad.net/or/+bug/1620802)

peternewell commented:
I have been working on a bug with the resume of the advanced adhesion -

http://www.elvastower.com/forums/index.php?/topic/35096-issue-with-advanced-adhesion-after-saveload/page__view__findpost__p__272082

There is an outstanding issue with electric locomotives that appears to be related to the resumption of the Power circuit. When initially resumed Power is indicated as being **Off**, and hence no MotiveForce is produced. (See screen shot of electric locomotive in the post above mine)

This issue seems to be more related to the work that you are doing on power circuits, and I don't want to interfere with the code related to your changes, hence could you have a look at it and correct it?

If you need any more details then please ask.

Thanks

Sharpe49 committed:
Fixed https://bugs.launchpad.net/or/+bug/1930124

YoRyan committed:
Merge branch 'master' into Sharpe49-power-supply

YoRyan committed:
chore(locales): update Orts.Simulation.pot

YoRyan committed:
docs: fix inconsistent titles

YoRyan committed:
docs: fix broken power supply scripting link

YoRyan commented on Source/ORTS.Settings/UserSettings.cs:
```diff
@@ -251,6 +251,8 @@ public enum DirectXFeature
         public bool OverrideNonElectrifiedRoutes { get; set; }
         [Default(true)]
         public bool HotStart { get; set; }
+        [Default(false)]
+        public bool NoDieselEngineStart { get; set; }
```
Tying the power mode to a global setting like this seems too inflexible; surely this should be a property of the selected consist and/or the selected activity. For example, imagine an activity that starts your locomotive in diesel mode in a non-electrified portion of track, with the expectation that you will change over to electric mode upon reaching the electrified section. The activity designer would want to ensure that the player's locomotive starts up in the correct mode.

You also have to consider other kinds of locomotives that can't be captured with a simple diesel on/diesel off toggle, such as dual overhead/third-rail units.

In the interest of getting this PR through, I would support keeping this setting for now, but I would like to move it to the Experimental tab. In the long-term, the goal should be to make it easy for the player to select which mode to start up in in explorer mode, and for activity and timetable designers to be able to select which modes they want their trains to start up in, thus allowing us to remove it.

Sharpe49 replied:
Ok to put it in the Experimental tab for the moment.
A parameter should be added to the timetables later.

YoRyan committed:
refactor: use briefer ternary operators

YoRyan commented on Source/Orts.Simulation/Common/Scripting/Common.cs:
```diff
@@ -90,6 +91,11 @@ public Timer(AbstractScriptClass asc)
             CurrentValue = asc.GameTime;
         }
+
+        public Timer(TrainCar car)
+        {
+            CurrentValue = () => (float)car.Simulator.GameTime;
+        }
```
When does any content script have access to the TrainCar class? It should be for internal use only.

Sharpe49 replied:
It is used for BatterySwitch and MasterKey classes which are not scriptable.
In the script, you don't have access to TrainCar objects anyway.

YoRyan commented on Source/RunActivity/Viewer3D/Viewer.cs:
```diff
@@ -547,14 +547,18 @@ public void SetCommandReceivers()
-                CircuitBreakerClosingOrderButtonCommand.Receiver = (MSTSElectricLocomotive)PlayerLocomotive;
-                CircuitBreakerOpeningOrderButtonCommand.Receiver = (MSTSElectricLocomotive)PlayerLocomotive;
-                CircuitBreakerClosingAuthorizationCommand.Receiver = (MSTSElectricLocomotive)PlayerLocomotive;
+                CircuitBreakerClosingOrderCommand.Receiver = (PlayerLocomotive as MSTSLocomotive).LocomotivePowerSupply;
+                CircuitBreakerClosingOrderButtonCommand.Receiver = (PlayerLocomotive as MSTSLocomotive).LocomotivePowerSupply;
+                CircuitBreakerOpeningOrderButtonCommand.Receiver = (PlayerLocomotive as MSTSLocomotive).LocomotivePowerSupply;
+                CircuitBreakerClosingAuthorizationCommand.Receiver = (PlayerLocomotive as MSTSLocomotive).LocomotivePowerSupply;
```
There's no benefit to the `as MSTSLocomotive` construction unless you also test for `null` when indexing into it. If the cast cannot be performed, both
```C#
(PlayerLocomotive as MSTSLocomotive).Xxx
```
and
```C#
(MSTSLocomotive)PlayerLocomotive.Xxx
```
will result in a runtime error.

Not going to change it, as this entire section needs to be refactored, really - just for future reference.

Sharpe49 replied:
I think that, in this case, this should be cleaner:
`if (PlayerLocomotive is MSTSElectricLocomotive electricLocomotive)` and use electricLocomotive in the code after.

About the runtime error, I don't think it can happen: we test that the locomotive is a MSTSElectricLocomotive and LocomotivePowerSupply is always initialized in the constructor of this class.

YoRyan committed:
refactor: centralize cab control power supply logic

YoRyan committed:
chore: remove unnecssary public setters

YoRyan commented on Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/ILocomotivePowerSupply.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace Orts.Simulation.RollingStocks.SubSystems.PowerSupplies
+{
+    public interface ILocomotivePowerSupply : IPowerSupply
```
It would be helpful if you included a quick summary of what this interface is useful for.

YoRyan commented on Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/IPowerSupply.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace Orts.Simulation.RollingStocks.SubSystems.PowerSupplies
+{
+    public interface IPowerSupply : ISubSystem<IPowerSupply>, IParsable
```
As above.

YoRyan commented on Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/DieselPowerSupply.cs:
```diff
@@ -0,0 +1,334 @@
+        private Timer PowerOnTimer;
+        private Timer AuxPowerOnTimer;
+
+        private DieselEngineState PreviousFirstEngineState;
+        private DieselEngineState PreviousSecondEngineState;
```
Why should we be limited to two diesel engines? Perhaps a dictionary would be the way to go here.

Sharpe49 replied:
We are only using these variables for the EnginePowerOn/Off and SecondEnginePowerOn/Off events.

YoRyan commented on Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/Pantograph.cs:
```diff
@@ -342,20 +347,30 @@ public void HandleEvent(PowerSupplyEvent evt)
                         }
+
+
+                        if (!Simulator.TRK.Tr_RouteFile.Electrified)
+                            Simulator.Confirmer.Warning(Simulator.Catalog.GetString("No power line!"));
```
I'm probably being ambitious here, but: If we ever get our own route editor, this logic may need to be changed. We can expect OR to support electrified and non-electrified sections on the same route.

Sharpe49 replied:
This code was moved from a MSTSElectricLocomotive function:
public override void SignalEvent(PowerSupplyEvent evt, int id) (function has been removed)

YoRyan commented on Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/Pantograph.cs:
```diff
@@ -342,20 +347,30 @@ public void HandleEvent(PowerSupplyEvent evt)
+
+                        if (!Simulator.TRK.Tr_RouteFile.Electrified)
+                            Simulator.Confirmer.Warning(Simulator.Catalog.GetString("No power line!"));
+                        if (Simulator.Settings.OverrideNonElectrifiedRoutes)
+                            Simulator.Confirmer.Information(Simulator.Catalog.GetString("Power line condition overridden."));
```
I think a warning like this is redundant. Players should already know that they're forcing electrification on.

Sharpe49 replied:
This code was moved from a MSTSElectricLocomotive function:
public override void SignalEvent(PowerSupplyEvent evt, int id) (function has been removed)

YoRyan commented on Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/PassengerCarPowerSupply.cs:
```diff
@@ -0,0 +1,452 @@
+        {
+            CarId = Train?.Cars.IndexOf(Wagon) ?? 0;
+
+            if (firstUpdate)
+            {
+                firstUpdate = false;
```
Testing for the first update like this is brittle and unlikely to survive a simulator save/restore. Could we possibly move this logic to Initialize()?

Sharpe49 replied:
We can't:
The Train member is not initialized when the Initialize is called.

YoRyan commented on Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/ElectricPowerSupply.cs:
```diff
@@ -15,65 +15,40 @@
-                return (float)Simulator.TRK.Tr_RouteFile.MaxLineVoltage;
-            }
-        }
+        public bool RouteElectrified => Simulator.TRK.Tr_RouteFile.Electrified || Simulator.Settings.OverrideNonElectrifiedRoutes;
+        public float LineVoltageV => (float)Simulator.TRK.Tr_RouteFile.MaxLineVoltage;
```
Again, assuming route-wide electrification is not particularly future-proof.

YoRyan commented on Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerSupplies/SteamPowerSupply.cs:
```diff
@@ -0,0 +1,129 @@
+    /// Basic power supply class for steam locomotives
+    /// For electrical systems powered by battery
+    /// </summary>
+    public class SteamPowerSupply : ILocomotivePowerSupply, ISubSystem<SteamPowerSupply>
```
So, no scriptable steam locomotive power supply? @peternewell would be disappointed. :)

Sharpe49 replied:
Do you really have a complex electric power supply system on steam locomotives?
I don't think so! :D

YoRyan reviewed (commented):
Right, this PR has sat in the queue long enough. I've performed a review and left some comments. Thanks!

YoRyan committed:
refactor: clarify use of the first update variable

YoRyan committed:
chore: clarify use of the engine state variables for sound triggers

YoRyan committed:
chore: add summaries for new power supply interfaces

YoRyan reviewed (approved):
Approved after a few more changes. LTGM. @cjakeman
